Added tests for weights.jl and misc.jl

Dawny33 opened the pull request:
Added tests for the following:

In the `weights.jl` tests:

- `weights` for `weights(vs::RealArray) = WeightVec(vec(vs))`
- `Base.sum(v::BitArray, w::WeightVec) = wsum(v, values(w))`
- `Base.sum(v::SparseMatrixCSC, w::WeightVec) = wsum(v, values(w))`
- `wsum(v::AbstractArray, w::AbstractVector) = dot(vec(v), w)`

In the `misc.jl` file:

- Added tests for `levelsmap`

nalimilan commented on test/misc.jl:
```diff
@@ -15,6 +15,13 @@ vals, lens = rle(z)
+y = [1, 1, 2, 2, 2, 3, 1, 2, 2, 3, 3, 3, 3]
+
+@test levelsmap(y) == Dict([(2, 2), (3, 3), (1, 1)])
+@test levelsmap(z) == Dict([(false, 2), (true, 1) ])
```
Probably better use a new vector here instead of `z`, which is redefined in each block above. It would be clearer (you can even use a `let` block around the two instructions you added).

Dawny33 replied:
I did had second thoughts about using `z`, but as I felt it'd be a nice example for the `levelsmap` implementation, I went ahead with it.

I'd use a different vector instead of it now.

nalimilan commented on test/weights.jl:
```diff
@@ -15,13 +17,33 @@ wv = weights(w)
+@test !isempty(bv)
+
+c  = [1, 4, 3, 5]; d = [4, 7, 18, 9]; e = [1, 2, -5, 3];
+s  = sparse(c, d, e)
```
What are `s`, `d` and `e` for?

Dawny33 replied:
`c`, `d` and `e` are vectors which I used for constructing the sparse matrix `s`.

nalimilan replied:
But `s` isn't used, right?

nalimilan commented on test/weights.jl:
```diff
@@ -15,13 +17,33 @@ wv = weights(w)
+@test eltype(cv) == Int64
+@test length(cv) == 4
+@test values(cv) == c
+@test sum(cv)    == 13.0
```
Better use `===` here and above, to check the return type too (should be `13` here, not `13.0`).

Dawny33 replied:
Yeah, `===` makes the test more robust. I would add it to [this line](https://github.com/JuliaStats/StatsBase.jl/pull/152#discussion-diff-56038439L15) too.

nalimilan commented:
Thanks!

Dawny33 commented:
^ Oops!  Working on it!

Dawny33 commented:
@nalimilan Pl review!

nalimilan commented on test/misc.jl:
```diff
@@ -15,6 +15,14 @@ vals, lens = rle(z)
+a = [1, 1, 2, 2, 2, 3, 1, 2, 2, 3, 3, 3, 3, 2]
+b = [true, false, false, true, false, true, true, false]
+
+@test levelsmap(a) == Dict([(2, 2), (3, 3), (1, 1)])
```
The best syntax for this is `Dict(2=>2, 3=>3, 1=>1)`.

nalimilan commented on test/weights.jl:
```diff
@@ -5,23 +5,45 @@ using Base.Test
+@test eltype(bv) == Bool
+@test length(bv) == 3
+@test values(bv) == b
+@test sum(bv)    == 3
```
Use `===` here too (you could even use it for all lines here).

nalimilan commented on test/weights.jl:
```diff
@@ -5,23 +5,45 @@ using Base.Test
 
 @test wsum(Float64[], Float64[]) === 0.0
 @test wsum(x, w) == 72.0
+@test wsum(p, q) == 29.0
```
Use `===` here and above.

Dawny33 commented:
@nalimilan Removed the sparse matrix test, as it doesn't solve the purpose.

Fixed the others by replacing the `==` with a `===` as suggested.

Changed the `Dict` format as suggested.

nalimilan commented:
Thanks! Other PRs like that would be welcome. ;-)

I'll leave it open for a few days to leave the occasion for others to comment, then I'll merge.

Dawny33 commented:
@nalimilan Sure!

Meanwhile, you can have a look at this too: https://github.com/JuliaStats/StatsBase.jl/pull/153
